### PR TITLE
stash: compact and consolidate on append

### PR DESCRIPTION
**Pull request by mjibson**
This allows append to clean up retracted data without users caring,
and allows peek to still work.

Item 3 in #12270.

### Motivation

  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Review by petrosagg — COMMENTED** [review]
The PR does what it says on the tin but it's unclear to me why we want to have atomic guarantees on compaction. Is there an explainer of why we need this?

File: src/stash/src/postgres.rs [hunk 1]
```diff
@@ -700,7 +726,13 @@ impl Append for Postgres {
                         .await?;
                     Self::seal_batch_tx(tx, std::iter::once((batch.collection_id, &batch.upper)))
                         .await?;
+                    Self::compact_batch_tx(
+                        tx,
+                        std::iter::once((batch.collection_id, &batch.compact)),
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/stash/src/postgres.rs:731]: Using both that and `&[..]` rust complains:

```
error[E0277]: `[(i64, &Antichain<i64>); 1]` is not an iterator
   --> src/stash/src/postgres.rs:729:48
    |
729 |                     Self::compact_batch_tx(tx, [(batch.collection_id, &batch.compact)]).await?;
    |                     ----------------------     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `[(i64, &Antichain<i64>); 1]` is not an iterator
    |                     |
    |                     required by a bound introduced by this call
    |
    = help: the trait `Iterator` is not implemented for `[(i64, &Antichain<i64>); 1]`
note: required by a bound in `Postgres::compact_batch_tx`
   --> src/stash/src/postgres.rs:319:12
    |
314 |     async fn compact_batch_tx<'a, I>(
    |              ---------------- required by a bound in this
...
319 |         I: Iterator<Item = (Id, &'a Antichain<Timestamp>)>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Postgres::compact_batch_tx`
```

[the remark this answers is not in the document]
Comment (mjibson) [reply src/stash/src/postgres.rs:731]: Making the bound `IntoIterator` breaks other stuff that's not worth fixing. I'm gonna leave this.

[the remark this answers is not in the document]
Comment (petrosagg) [reply src/stash/src/postgres.rs:731]: ah, ok. I thought the receiving function was expecting an `IntoIterator`, nevermind

**Review by petrosagg — APPROVED** [review]

**Pull request merged**